### PR TITLE
Git Info Logging

### PR DESCRIPTION
**Pull request by cSchubes**
Adds the following features:

- Git utility module to determine if a directory is inside a repo and obtain the SHA, tracked file patch, and untracked file patch. Also searches for `git_info.json` in a file tree, which is used to find git info stored by packages in `site-packages` after install.
- Log git SHA, tracked file patch, and untracked file patch for `ravenML` on every dataset creation or model training call. 
- Log git SHA, tracked file patch, and untracked file patch for dataset plugin used for dataset creation if possible. 
- Log git SHA, tracked file patch, and untracked file patch for training plugin used for model training if possible. 

**Caveat:** Logging git information only works if the package was installed locally (`pip install -e .`), from PyPI (`pip install ravenml`) or from GitHub using a clone (`pip install "https://github.com/autognc/ravenML#egg=ravenml"`). It does NOT work if the package was installed using a GitHub tarball (`pip install "https://github.com/autognc/ravenML/tarball/master"`). This is because in that case, git info is not available at install time nor at runtime. In the case of local install, git info is available at runtime. In the case of PyPI/git clone install, git info is available at install time.

When doing a PyPI or git clone install, the `ravenML` package now creates a file called `git_info.json` and stores it as a data file alongside the package source code installed at `site-packages`.

An upcoming PR will include zipping/tarring all source code and storing alongside model/dataset artifacts to ensure repeatability even when a tarball install is used.

### CODE REVIEW

**Comment by cSchubes** [comment]
sorry for so many commits, I had to test when doing tarball/git clone installs

File: ravenml/utils/git.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+Utility functions for grabbing git information to place into metadata.
+"""
+
+import subprocess
+import json
+from os import chdir, listdir
+from pathlib import Path
+
+def is_repo(path: Path) -> bool:
```
Comment (kvablack) [inline ravenml/utils/git.py]: Type hint shouldn't be bool

**Review by kvablack — DISMISSED** [review]
LGTM! This looks a lot more complicated than I was expecting, your solutions are genius. Only thought I have is about the readability of storing git patches as one giant JSON entry, it might be easier on the user if they were their own file. Also, is it possible to combine tracked and untracked patches, since really all we care about are any changes from the latest commit? Not a big deal though, as long as the info is there, just maybe some things to look at when doing the full source code zipping.

**Comment by cSchubes** [comment]
> LGTM! This looks a lot more complicated than I was expecting, your solutions are genius. Only thought I have is about the readability of storing git patches as one giant JSON entry, it might be easier on the user if they were their own file. Also, is it possible to combine tracked and untracked patches, since really all we care about are any changes from the latest commit? Not a big deal though, as long as the info is there, just maybe some things to look at when doing the full source code zipping.

I think leaving it as JSON is worth it so we can put them in the metadata.json file. Keeping the number of individual files at a minimum seems best to me but I'm open to discussion on that.

As far as I could tell there wasn't a good way to combine the patches and still be able to use the patchfile. agreed its kind of a pain. 

can you explicitly approve the PR?

**Review by kvablack — APPROVED** [review]

**Pull request merged**